### PR TITLE
Rework parameters passing for functions within `ec.rebuild`

### PR DESCRIPTION
**Pull request by ghost**
# What problem are we solving?

https://github.com/seaweedfs/seaweedfs/issues/7416

# How are we solving the problem?

Simplifies the overall codebase and allows to cleanly handle parallelization via waitgroups.

# How is the PR tested?

No functional changes.

# Checks
- [ ] I have added unit tests if possible.
- [ ] I will add related wiki document changes and link to this PR after merging.

## Summary by CodeRabbit

* **Refactor**
  * Reorganized EC rebuild into a single stateful rebuilder to centralize control flow, node selection, disk-space checks, messaging, and per-volume processing for more reliable recoveries.

* **Tests**
  * Added comprehensive unit tests validating shard distribution, node selection heuristics, space/error paths, duplicate shard handling, and multiple rebuild edge cases.

### CODE REVIEW

File: weed/shell/command_ec_rebuild.go [hunk 1]
```diff
@@ -208,29 +246,28 @@ func prepareDataToRecover(commandEnv *CommandEnv, rebuilder *EcNode, collection
...
+			copyErr = operation.WithVolumeServerClient(false, pb.NewServerAddressFromDataNode(rebuilder.info), erb.commandEnv.option.GrpcDialOption, func(volumeServerClient volume_server_pb.VolumeServerClient) error {
 				_, copyErr := volumeServerClient.VolumeEcShardsCopy(context.Background(), &volume_server_pb.VolumeEcShardsCopyRequest{
 					VolumeId:       uint32(volumeId),
 					Collection:     collection,
 					ShardIds:       []uint32{uint32(shardId)},
 					CopyEcxFile:    needEcxFile,
 					CopyEcjFile:    true,
 					CopyVifFile:    needEcxFile,
-					SourceDataNode: ecNodes[0].info.Id,
+					SourceDataNode: rebuilder.info.Id,
```
[the remark this answers is not in the document]
Comment (ghost) [reply weed/shell/command_ec_rebuild.go]: Not a bug - `ecNodes` is no longer pre-sorted, so `rebuilder` is resolved via `ecNodeWithMoreFreeSlots()`.

File: weed/shell/command_ec_rebuild.go [hunk 2]
```diff
@@ -239,9 +276,9 @@ func prepareDataToRecover(commandEnv *CommandEnv, rebuilder *EcNode, collection
 			}
 		}
 		if copyErr != nil {
-			fmt.Fprintf(writer, "%s failed to copy %d.%d from %s: %v\n", rebuilder.info.Id, volumeId, shardId, ecNodes[0].info.Id, copyErr)
+			erb.write("%s failed to copy %d.%d from %s: %v\n", rebuilder.info.Id, volumeId, shardId, rebuilder.info.Id, copyErr)
 		} else {
-			fmt.Fprintf(writer, "%s copied %d.%d from %s\n", rebuilder.info.Id, volumeId, shardId, ecNodes[0].info.Id)
+			erb.write("%s copied %d.%d from %s\n", rebuilder.info.Id, volumeId, shardId, rebuilder.info.Id)
 			copiedShardIds = append(copiedShardIds, uint32(shardId))
 		}
```
[the remark this answers is not in the document]
Comment (ghost) [reply weed/shell/command_ec_rebuild.go:290]: Not a bug - `ecNodes` is no longer pre-sorted, so `rebuilder` is resolved via `ecNodeWithMoreFreeSlots()`.
